core/async/src/test_loop/test_loop_sender.rs

```diff
@@ -0,0 +1,155 @@
+// This was specifically required as rust was complaining the traits Sync/Send are not implemented for actor A
+unsafe impl<A> Send for TestLoopSender<A> {}
+unsafe impl<A> Sync for TestLoopSender<A> {}
+
+impl<A> Clone for TestLoopSender<A> {
+    fn clone(&self) -> Self {
+        Self {
+            actor: self.actor.clone(),
+            pending_events_sender: self.pending_events_sender.clone(),
+            shutting_down: self.shutting_down.clone(),
+        }
+    }
+}
+
+impl<A> DelayedActionRunner<A> for TestLoopSender<A>
+where
+    A: 'static,
+{
+    fn run_later_boxed(
+        &mut self,
+        name: &str,
+        dur: Duration,
+        f: Box<dyn FnOnce(&mut A, &mut dyn DelayedActionRunner<A>) + Send + 'static>,
+    ) {
+        if self.shutting_down.load(Ordering::Relaxed) {
+            return;
+        }
+
+        let mut this = self.clone();
+        let callback = move || f(&mut this.clone().actor.lock().unwrap(), &mut this);
+        self.pending_events_sender.send_with_delay(
+            CallbackEvent {
+                description: format!("DelayedActionRunner {:?}", name),
+                callback: Box::new(callback),
+            },
+            dur,
+        );
+    }
+}
+
+impl<M, A> CanSend<M> for TestLoopSender<A>
+where
+    M: actix::Message + Send + 'static,
+    A: Actor + HandlerWithContext<M> + 'static,
+{
+    fn send(&self, msg: M) {
+        let this = self.clone();
+        let callback = move || {
+            this.actor.lock().unwrap().handle(msg, &mut this.clone());
+        };
+        self.pending_events_sender.send_with_delay(
+            CallbackEvent {
+                description: format!("Message {:?}", type_name::<M>()),
+                callback: Box::new(callback),
+            },
+            Duration::ZERO,
+        );
+    }
+}
+
+impl<M, R, A> CanSend<MessageWithCallback<M, R>> for TestLoopSender<A>
+where
+    M: actix::Message<Result = R> + Send + 'static,
+    A: Actor + HandlerWithContext<M> + 'static,
+    R: 'static,
+{
+    fn send(&self, msg: MessageWithCallback<M, R>) {
+        let this = self.clone();
+        let callback = move || {
+            let MessageWithCallback { message: msg, callback } = msg;
+            let result = this.actor.lock().unwrap().handle(msg, &mut this.clone());
+            callback(Ok(result));
+        };
+        self.pending_events_sender.send_with_delay(
+            CallbackEvent {
+                description: format!("Message {:?}", type_name::<M>()),
+                callback: Box::new(callback),
+            },
+            Duration::ZERO,
+        );
+    }
+}
+
+impl<A> TestLoopSender<A>
+where
+    A: Actor + 'static,
+{
+    pub fn new(
+        actor: A,
+        pending_events_sender: DelaySender<CallbackEvent>,
+        shutting_down: Arc<AtomicBool>,
+    ) -> Self {
+        Self { actor: Arc::new(Mutex::new(actor)), pending_events_sender, shutting_down }
```

Comment:
why is Mutex needed here if the loop is running single threaded?

Reply:
Sorta answered [here](https://github.com/near/nearcore/pull/11450#discussion_r1622972876), we need rust to be satisfied that all interfaces match.